src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -771,7 +841,7 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
 
         private void ResetRequestHeaderParsingState()
         {
-            if (_requestHeaderParsingState != RequestHeaderParsingState.Trailers)
+            if (_requestHeaderParsingState != RequestHeaderParsingState.Trailers && _stateManager.State != ConnectionState.Closed)
             {
                 _highestOpenedStreamId = _currentHeadersStream.StreamId;
```

Comment:
Hmm there's still a race here.

Reply:
Might be easier to use locks instead of Interlocked since otherwise we need to combine state and highest stream id into a single field.

Reply:
:+1: for using locks for connection state transitions. I don't suspect these will be high contention.